### PR TITLE
Allow for ref_model=None in DPOTrainer

### PR DESCRIPTION
**Pull request by vincentmin**
PR for #624. The goal is to allow DPOTrainer to allow for `ref_model=None`. If the model to be trained is a `peft` model, or a `peft_config` is provided, we use the base model for improved memory consumption. Otherwise, a reference model is created on the fly by copying the `model`.

### CODE REVIEW

**Comment by vincentmin** [comment]
@lvwerra I used a bit of a different approach from the `PPOTrainer`. In `PPOTrainer` `self.ref_model` can be None in which case `self.accelerator.unwrap_model(self.model).pretrained_model.disable_adapter()` is used. Instead, I just set `self.ref_model=model.get_base_model()`. Let me know if you see any problems with this approach.

**Comment by kashif** [comment]
will do! thanks! i wanted to test it out together with the PR #639

**Review by younesbelkada — COMMENTED** [review]
Great work @vincentmin for adding this support, there is a small merge conflict, would you be happy to take a look? Otherwise happy to do it!

**Comment by vincentmin** [comment]
@younesbelkada Thanks. The conflict is resolved.

**Review by younesbelkada — COMMENTED** [review]
Hmm a CI test is failing, you might need to add a simple check here: https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L158 to make the CI happy

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -155,13 +167,13 @@ def __init__(
 
         if disable_dropout:
             disable_dropout_in_model(model)
-            disable_dropout_in_model(ref_model)
+            if self.ref_model:
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:170]:
```suggestion
            if self.ref_model is not None:
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot ! Left a last suggestion, what do you think?

**Comment by tannonk** [comment]
@vincentmin Thanks for the effort on this PR! Any chance you could provide the script/run command you used to test the implementation? I'm yet to get this working from my end.

**Comment by vincentmin** [comment]
If you modify [this script](https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama_2/scripts/dpo_llama2.py) by setting `model_ref=None` on line 141, it should work. I haven't had a chance to try this myself, so let me know if you encounter an error.

**Comment by vincentmin** [comment]
@lvwerra @younesbelkada Could you give your input here? @tannonk 's proposed fix sounds reasonable to me, but I do not have the time to verify it at the moment. If this is indeed a correct fix, I would expect this to be necessary for the PPOTrainer as well.

**Comment by kashif** [comment]
thanks @tannonk I can test it out

**Pull request merged**